README.md
# vFlightD3ck #

A deck management and planning tool for simulated aircraft carrier operations.

Technically, a virtual [Ouja board](https://commons.wikimedia.org/wiki/Category:Ouija_boards_(aircraft_carrier)).

## Installation
This app can be hosted on any webserver.
Download the latest release from the [releases page] and unpack it in the server HTTP documents folder of your choice.
It can also be hosted locally with something like [Portable Apache](https://apr.apache.org/).

## Usage
See the [user manual](doc/MANUAL.md)

## Documentation
Click the "Help" button to open the user manual in a new browser tab.

If you want to add other decks or aircraft shapes to this tool,
see [OBJECTS.md](doc/OBJECTS.md)

If you are interested in the source code of this tool,
have a look at the [DEVELOP.md](doc/DEVELOP.md)` document.

#### Dependencies
vFlightD3ck depends on these external libraries, included with the code:

- [D3JS v4](https://d3js.org/)
- [jQuery v3.2.1](https://jquery.com/)

## Feedback and contacts
If you think there is a bug in the program, or something is missing or wrong
with the documentation/support files, feel free to [open an issue].

## License
This software is released under the LGPL V3 license.
A copy is included in the LICENSE file.

[open an issue]: issues
[releases page]: releases

changelog.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Added
- Nothing
### Changed
- Nothing
### Deprecated
- Nothing
### Removed
- Nothing
### Fixed
- Nothing
### Security
- Nothing

## 0.1.0 - 2018-06-09
First prototype


[Unreleased]: https://github.com/marcosox/project/compare/v0.1.0...HEAD


js/api.js
/*
 * Created by sox on 02/07/2017.
 */

// ******************************** aircraft API functions

function moveAircraftToHotspot(aircraft, hotspot) {
	setAircraftPosition(aircraft, hotspot.x, hotspot.y);
	setAircraftRotation(aircraft, hotspot.r);
}

function setAircraftShape(aircraft, shape) {
	aircraft.shape = shape;
}

function setAircraftColor(aircraft, color) {
	aircraft.fill = color;
}

/**
 * changes the label of the specified aircraft
 * @param aircraft
 * @param label
 */
function setAircraftLabel(aircraft, label) {
	aircraft.label = label;
}

/**
 *
 * @param aircraft aircraft object in d3.data()
 * @param dx
 * @param dy
 */
function moveAircraft(aircraft, dx, dy) {
	aircraft.x += dx;
	aircraft.y += dy;
}

/**
 * move aircraft to absolute position
 * @param aircraft aircraft object in d3.data()
 * @param x
 * @param y
 */
function setAircraftPosition(aircraft, x, y) {
	aircraft.x = x;
	aircraft.y = y;
}

/**
 * Rotate an aircraft to a relative value
 * @param aircraft aircraft object in d3.data()
 * @param rotationDelta relative rotation in degrees
 */
function rotateAircraft(aircraft, rotationDelta) {
	aircraft.r += rotationDelta;
}

/**
 * rotate an aircraft to an absolute rotation
 * @param aircraft aircraft object in d3.data()
 * @param rotation absolute rotation in degrees
 */
function setAircraftRotation(aircraft, rotation) {
	aircraft.r = rotation;
}


// **************************************************** decks API functions

/**
 * imports decks from serialized form (e.g. from file json)
 * configs are in the form {deckId, config}[]
 * @param configs
 */
function importDeckConfigs(configs) {
	configs.forEach(function (configObj) {
		decks[configObj.deckId].config = configObj.config;
	});
}

function clearDeckConfigs(decksKeys) {
	decksKeys.forEach(function (key) {
		decks[key].config = [];
	});
}

function exportCurrentDeckConfig() {
	return exportDeckConfigs([activeDeckId]);
}

function exportAllDeckConfigs() {
	return exportDeckConfigs(Object.keys(decks));
}

/**
 * exports an array of decks into serializable form (the one used by download)
 * @param decksIDs
 * @returns {Array}
 */
function exportDeckConfigs(decksIDs) {
	var result = [];
	decksIDs.forEach(function (key) {
		result.push({
			deckId: key,
			config: decks[key].config
		});
	});
	return result;
}


js/installedAircrafts.js
/*
 * Created by sox on 02/07/2017.
 */

var installedAircrafts = {
	F18: {
		title: "F18 Hornet",
		heigth: 65,
		width: 39,
		defaultX: 0.20,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"fold", "spread"]
	},
	C2: {
		title: "C2 Greyhound",
		heigth: 66,
		width: 86,
		defaultX: 0.55,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"fold", "spread"]
	},
	E2: {
		title: "E2 Hawkeye",
		heigth: 67,
		width: 87,
		defaultX: 0.45,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"fold", "spread"]
	},
	EA6: {
		title: "EA6B Prowler",
		heigth: 67,
		width: 60,
		defaultX: 0.40,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"fold", "spread"]
	},
	SH60: {
		title: "SH60 Seahawk",
		heigth: 56,
		width: 56,
		defaultX: 0.60,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"fold", "spread", "running"]
	},
	EH101: {
		title: "EH101 Merlin",
		heigth: 79,
		width: 67,
		defaultX: 0.30,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"fold", "spread", "running"]
	},
	F35: {
		title: "F35 Lightning",
		heigth: 59,
		width: 49,
		defaultX: 0.65,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"fold", "spread"]
	},
	F14: {
		title: "F14 Tomcat",
		heigth: 72,
		width: 72,
		defaultX: 0.57,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"fold", "spread"]
	},
	T45: {
		title: "T45 Goshawk",
		heigth: 45,
		width: 37,
		defaultX: 0.35,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"spread"]
	},
	AV8: {
		title: "AV8B Harrier",
		heigth: 63,
		width: 35,
		defaultX: 0.70,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"spread"]
	},
	RAFALE: {
		title: "Dassault Rafale",
		heigth: 58,
		width: 41,
		defaultX: 0.73,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"spread"]
	},
	CH53: {
		title: "MH53 Sea Dragon",
		heigth: 97,
		width: 80,
		defaultX: 0.37,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"fold", "spread", "running"]
	},
	AB212: {
		title: "AB212 Twin Huey",
		heigth: 60,
		width: 60,
		defaultX: 0.27,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"spread", "running"]
	},
	NH90: {
		title: "NH90",
		heigth: 76,
		width: 65,
		defaultX: 0.23,
		defaultY: 200,
		fontSize: 15,
		shapes: [
			"fold", "spread", "running"]
	},
	CRANE: {
		title: "Large crash crane",
		heigth: 56,
		width: 20,
		defaultX: 0.15,
		defaultY: 200,
		fontSize: 8,
		shapes: [
			"normal"]
	},
	FORKLIFT_SMALL: {
		title: "Small forklift",
		heigth: 20,
		width: 9,
		defaultX: 0.16,
		defaultY: 200,
		fontSize: 8,
		shapes: [
			"normal"]
	},
	FORKLIFT_LARGE: {
		title: "Large forklift",
		heigth: 25,
		width: 9,
		defaultX: 0.17,
		defaultY: 200,
		fontSize: 8,
		shapes: [
			"normal"]
	},
	TOWBAR: {
		title: "Tow bar",
		heigth: 15,
		width: 9,
		defaultX: 0.18,
		defaultY: 200,
		fontSize: 8,
		shapes: [
			"normal"]
	},
	TOWTRACTOR: {
		title: "Tow/Fire/Oxygen/Crash tractor",
		heigth: 20,
		width: 10,
		defaultX: 0.19,
		defaultY: 200,
		fontSize: 8,
		shapes: [
			"normal"]
	}
};

js/installedDecks.js
/*
 * Created by sox on 29/06/2017.
 */
var installedDecks = {
	nimitzDeck: {
		image: "nimitz_deck.png",
		title: "CVN68 USS Nimitz - Flight Deck",
		hotspots: [
			{
				title: 'CAT 1',
				x: 848,
				y: 260,
				r: 85
			}, {
				title: 'CAT 2',
				x: 827,
				y: 184,
				r: 89
			}, {
				title: 'CAT 3',
				x: 470,
				y: 124,
				r: 86
			}, {
				title: 'CAT 4',
				x: 414,
				y: 77,
				r: 90
			}]
	},
	nimitzDeckSimple: {
		image: "nimitz_deck_simple.png",
		title: "CVN68 USS Nimitz - Flight Deck (no details)",
		hotspots: [
			{
				title: 'CAT 1',
				x: 848,
				y: 260,
				r: 85
			}, {
				title: 'CAT 2',
				x: 827,
				y: 184,
				r: 89
			}, {
				title: 'CAT 3',
				x: 470,
				y: 124,
				r: 86
			}, {
				title: 'CAT 4',
				x: 414,
				y: 77,
				r: 90
			}]
	},
	nimitzHangar: {
		image: "nimitz_hangar.png",
		title: "CVN68 USS Nimitz - Hangar Bay",
		hotspots: []
	},
	nimitzDual: {
		image: "nimitz_dual.png",
		title: "CVN68 USS Nimitz - Dual Deck view",
		hotspots: [
			{
				title: 'CAT 1',
				x: 838,
				y: 217,
				r: 85
			}, {
				title: 'CAT 2',
				x: 830,
				y: 142,
				r: 89
			}, {
				title: 'CAT 3',
				x: 469,
				y: 80,
				r: 86
			}, {
				title: 'CAT 4',
				x: 401,
				y: 33,
				r: 90
			}]
	},
	nimitzDualSimple: {
		image: "nimitz_dual_simple.png",
		title: "CVN68 USS Nimitz - Dual Deck view (no details)",
		hotspots: [
			{
				title: 'CAT 1',
				x: 845,
				y: 200,
				r: 85
			}, {
				title: 'CAT 2',
				x: 828,
				y: 124,
				r: 89
			}, {
				title: 'CAT 3',
				x: 471,
				y: 63,
				r: 86
			}, {
				title: 'CAT 4',
				x: 403,
				y: 17,
				r: 90
			}]
	},
	garibaldi: {
		image: "garibaldi.png",
		title: "C551 Garibaldi - Flight Deck",
		hotspots: [
			{
				title: "Elevator 1",
				x: 498,
				y: 221,
				r: 90
			},
			{
				title: "Elevator 2",
				x: 814,
				y: 223,
				r: 270
			}, {
				title: "Takeoff position 1",
				x: 496,
				y: 168,
				r: 92
			}, {
				title: "Takeoff position 2",
				x: 395,
				y: 166,
				r: 92
			}
		]
	},
	cavourDeck: {
		image: "Cavour_deck.png",
		title: "C550 Cavour - Flight Deck",
		hotspots: [
			{
				title: "Elevator 1",
				x: 420,
				y: 287,
				r: 0
			}, {
				title: "Elevator 2",
				x: 775,
				y: 218,
				r: 270
			}, {
				title: "Takeoff position 1",
				x: 536,
				y: 154,
				r: 90
			}, {
				title: "Takeoff position 2",
				x: 419,
				y: 156,
				r: 90
			}]
	},
	cavourHangar: {
		image: "Cavour_hangar.png",
		title: "C550 Cavour - Hangar Bay",
		hotspots: [{
			title: "Elevator 2",
			x: 775,
			y: 218,
			r: 270
		}]
	},
	cavourDual: {
		image: "Cavour_dual.png",
		title: "C550 Cavour - Dual deck view",
		hotspots: [
			{
				title: "Elevator 1 (flight deck)",
				x: 411,
				y: 177,
				r: 0
			}, {
				title: "Elevator 2 (flight deck)",
				x: 769,
				y: 112,
				r: 270
			}, {
				title: "Takeoff position 1",
				x: 536,
				y: 46,
				r: 90
			}, {
				title: "Takeoff position 2",
				x: 419,
				y: 48,
				r: 90
			}, {
				title: "Elevator 2 (hangar deck)",
				x: 759,
				y: 314,
				r: 270
			}, {
				title: "Elevator 1 (hangar deck)",
				x: 406,
				y: 381,
				r: 0
			}]
	},
	USSWasp: {
		image: "USS_Wasp.png",
		title: "LHD USS Wasp - Flight Deck",
		hotspots: [
			{
				title: "Elevator 1",
				x: 225,
				y: 296,
				r: 0
			}, {
				title: "Elevator 2",
				x: 417,
				y: 103,
				r: 180
			}]
	},
	queenElizabeth: {
		image: "Queen_elizabeth.png",
		title: "HMS Queen Elizabeth - Flight Deck",
		hotspots: []
	},
	clemenceauDeck: {
		image: "Clemenceau_deck.png",
		title: "RS98 Clemenceau - Flight Deck",
		hotspots: [
			{
				title: 'CAT 1',
				x: 904,
				y: 148,
				r: 88
			}, {
				title: 'CAT 2',
				x: 561,
				y: 116,
				r: 80
			}]
	},
	cdgDeck: {
		image: "CharlesDeGaulle_deck.png",
		title: "R91 Charles de Gaulle - Flight Deck",
		hotspots: [
			{
				title: 'CAT 1',
				x: 792,
				y: 171,
				r: 90
			}, {
				title: 'CAT 2',
				x: 544,
				y: 124,
				r: 86
			}]
	},
	cdgHangar: {
		image: "CharlesDeGaulle_hangar.png",
		title: "R91 Charles de Gaulle - Hangar Bay",
		hotspots: []
	},
	cdgDual: {
		image: "CharlesDeGaulle_dual.png",
		title: "R91 Charles de Gaulle - Dual Deck view",
		hotspots: [
			{
				title: 'CAT 1',
				x: 792,
				y: 141,
				r: 90
			}, {
				title: 'CAT 2',
				x: 545,
				y: 94,
				r: 86
			}
		]
	},
	caseI: {
		image: "day-pattern.png",
		title: "CASE I pattern",
		hotspots: [
			{
				title: 'Commencing',
				x: 1163,
				y: 379,
				r: 90
			}, {
				title: 'Initial',
				x: 1031,
				y: 58,
				r: 270
			}, {
				title: 'Break',
				x: 438,
				y: 58,
				r: 270
			}, {
				title: 'Abeam',
				x: 738,
				y: 236,
				r: 90
			}, {
				title: 'Ball',
				x: 814,
				y: 100,
				r: 261
			}]
	},
	caseIII: {
		image: "CASE_III.png",
		title: "CASE III pattern",
		hotspots: [
			{
				title: 'Marshall',
				x: 1268,
				y: 97,
				r: 270
			}, {
				title: 'Platform',
				x: 1004,
				y: 214,
				r: 245
			}, {
				title: 'Dirty up',
				x: 794,
				y: 277,
				r: 270
			}, {
				title: 'Ball',
				x: 415,
				y: 347,
				r: 235
			}, {
				title: 'Tank high',
				x: 393,
				y: 42,
				r: 270
			}, {
				title: 'Tank low',
				x: 267,
				y: 142,
				r: 270
			}]
	},
	ccz: {
		image: "Radar_Screen.png",
		title: "CCZ Radar map",
		hotspots: []
	}
};